app/assets/javascripts/discourse-markdown-it/src/setup.js

```diff
@@ -0,0 +1,360 @@
+import deprecated from "discourse-common/lib/deprecated";
+import { cloneJSON } from "discourse-common/lib/object";
+import { textReplace } from "pretty-text/text-replace";
+import makeEngine, { cook } from "./engine";
+
+// note, this will mutate options due to the way the API is designed
+// may need a refactor
+export default function setupIt(features, options, siteSettings, state) {
+  Setup.run(features, options, siteSettings, state);
+}
+
+class Setup {
+  static run(features, options, siteSettings, state) {
+    if (options.setup) {
+      // Already setup
+      return;
+    }
+
+    const setup = new Setup(options);
+
+    features.sort((a, b) => a.priority - b.priority);
+
+    for (const feature of features) {
+      setup.#setupFeature(feature.id, feature.setup);
+    }
+
+    for (const entry of Object.entries(state.allowListed ?? {})) {
+      setup.allowList(entry);
+    }
+
+    setup.#runOptionsCallbacks(siteSettings, state);
+
+    setup.#enableMarkdownFeatures();
+
+    setup.#finalizeGetOptions(siteSettings);
+
+    setup.#makeEngine();
+
+    setup.#buildCookFunctions();
+  }
+
+  #context;
+  #options;
+
+  #allowListed = [];
+  #customMarkdownCookFunctionCallbacks = [];
+  #loadedFeatures = [];
+  #optionCallbacks = [];
+  #pluginCallbacks = [];
+
+  constructor(options) {
+    options.markdownIt = true;
+
+    this.#options = options;
+
+    // hack to allow moving of getOptions – see #finalizeGetOptions
+    this.#context = { options };
+  }
+
+  allowList(entry) {
+    this.#allowListed.push(entry);
+  }
+
+  registerOptions(entry) {
+    this.#optionCallbacks.push(entry);
+  }
+
+  registerPlugin(entry) {
+    this.#pluginCallbacks.push(entry);
+  }
+
+  buildCookFunction(entry) {
+    this.#customMarkdownCookFunctionCallbacks.push(entry);
+  }
+
+  #setupFeature(featureName, callback) {
+    // When we provide the API object to the setup callback, we expect them to
+    // make use of it synchronously. However, it is possible that the could
+    // close over the API object, intentionally or unintentionally, and cause
+    // memory leaks or unexpectedly call API methods at a later time with
+    // unpredictable results. This make sure to "gut" the API object after the
+    // callback is executed so that it cannot leak memory or be used later.
```

Comment:
This part is new and worth pointing out, we may want to do something like this in more places